src/reconcile/rec_visibility.c

```diff
@@ -357,20 +357,23 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
         }
         if (upd != NULL) {
             /*
-             * The beginning of the validity window is the selected update's time pair.
-             *
-             * FIXME-PM-1521: We shouldn't need to check for this. We're actually allowed to commit
-             * updates to the same key out of timestamp order. So we can have stop time pairs
-             * earlier than their respective start time pair. We need to figure out what to do in
-             * WT-5469.
+             * If we're seeing a non-timestamped tombstone being applied on top of a timestamped
+             * update, force the tombstone to be globally visible so that we destroy the key. This
+             * isn't technically correct but when we mix timestamps, we're not guaranteeing that
+             * that older readers will be able to continue reading content that has been made
+             * invisible by a non-timestamped update.
              */
-            if (upd->start_ts <= upd_select->stop_ts && upd->txnid <= upd_select->stop_txn) {
-                upd_select->durable_ts = upd_select->start_ts = upd->start_ts;
-                /* If durable timestamp is provided, use it. */
-                if (upd->durable_ts != WT_TS_NONE)
-                    upd_select->durable_ts = upd->durable_ts;
-                upd_select->start_txn = upd->txnid;
-            }
+            if (upd_select->stop_ts == WT_TS_NONE && upd->start_ts != WT_TS_NONE)
```

Comment:
I think this is the only part that has been added since most of you LGTM'd. If we're following up a timestamped update with a non-timestamped tombstone, I'm going to just ensure that we delete the key entirely. What I don't want is a malformed cell with a lower stop than the start. This is also not 100% correct in the same way that blowing away history content for a key isn't right but should be fine for MongoDB's purposes. In the long term, they should be able to remove the use of non-timestamped updates so this should be a temporary fix.

There's one more missing piece to this that I noticed while working on this section. If we delete a key due to a tombstone being globally visible, we need to destroy the history store content associated with that key rather than waiting for checkpoint. I've written up a scenario where we have visibility issues in the ticket: WT-5641.